docs: add guide for the case of high memory usage (#15042)
supabase/supabase: apps/docs/pages/guides/platform/exhaust-ram.mdx

import Layout from '~/layouts/DefaultGuideLayout'

export const meta = {
  id: 'exhaust-ram',
  title: 'High Memory Usage',
  description:
    'Learn what high RAM usage could mean for your Supabase instance and what could have caused it.',
}

Learn what high RAM usage could mean for your Supabase instance and what could have caused it.

## Base Memory Usage

> Besides Postgres, a wide range of services are running under the hood. These services take up memory even when your database is idle. Specifically on the smallest compute instance that comes with 1 GB of RAM, it is not unusual for your project to have a base memory usage of ~50%.

Reviewer: ```suggestion
You might observe elevated memory usage, even with little to no load. Besides Postgres, a wide range of services is running under the hood resulting in an elevated base memory usage. Especially on the smallest compute instance that comes with 1 GB of RAM, it is not unusual for your project to have a base memory usage of ~50%.
```

## Issues with High Memory Usage

Every Supabase project runs in its dedicated virtual machine. Your instance will have a different set of hardware provisioned depending on your [compute add-on](https://supabase.com/docs/guides/platform/compute-add-ons). Your hardware may not be suitable for the intended workload and may experience high RAM usage.

High RAM usage could come with a range of issues:

- degraded performance overall as your instance might have to rely more on swap
- the operating system might start killing processes as your system runs out of memory
- in rare cases, your instance may become unresponsive

## Monitor your RAM

> You can check your RAM usage on the Supabase Platform. For this go to database health in the reports section or [click here](https://app.supabase.com/project/_/reports/database) and select your project.

Reviewer: ```suggestion
To check your RAM usage on the Supabase Platform, head over to [database health in the reports section](https://app.supabase.com/project/_/reports/database).
```

Personal pref, feel free to ignore

<Admonition type="note">

High memory usage doesn't necessarily mean that your instance is at risk. Memory that is used for caching and buffers improves data access speed. But if you notice less performance alongside high memory usage, your memory usage might be unhealthy.

> </Admonition>

Reviewer 2: Wondering if it would be worth moving this higher up - possibly under the base memory usage (so that customers know high base usage can be okay)?

Author: Yes, good point. This is probably something they should know about first. Will move it even above the base memory usage 👍

<img width="885" alt="image" src="https://github.com/supabase/supabase/assets/1732217/5efbf668-f047-40a6-bb18-8be4c935174a">

> It is also possible to monitor your resources and set up alerts using Prometheus/Grafana. With Grafana you will be able to see how much of your RAM is used for caching. Moreover, you can track other metrics such as your Swap usage. You can find a guide [here](https://supabase.com/docs/guides/platform/metrics).

Reviewer 2: ```suggestion
It is also possible to monitor your resources and set up alerts using Prometheus/Grafana. With Grafana you will be able to see how much of your RAM is used for caching and you can track other metrics such as your Swap usage. Read the [Metrics Guide](https://supabase.com/docs/guides/platform/metrics) to learn more.
```

> It is also possible to monitor your resources and set up alerts using Prometheus/Grafana. With Grafana you will be able to see how much of your RAM is used for caching. Moreover, you can track other metrics such as your Swap usage. You can find a guide for this [here](https://supabase.com/docs/guides/platform/metrics).

Reviewer: ```suggestion
It is also possible to monitor your resources and set up alerts using Prometheus/Grafana. With Grafana you will be able to see how much of your RAM is used for caching. Moreover, you can track other metrics such as your Swap usage. You can find a guide [here](https://supabase.com/docs/guides/platform/metrics).
```

Just a nit, personal pref

## Common reasons for high RAM usage

Everything you do with your Supabase project requires memory in some form. Hence, there can be many reasons for high RAM usage. Here are some common ones:

- **Query performance:** Queries that take a long time to complete (>1 second) could be using your RAM inefficiently. Check our guide on [examining query performance](https://supabase.com/docs/guides/platform/performance#examining-query-performance).

> - **Too many connections:** Every connection to your database consumes memory. You can check the number of active connections [here](https://supabase.com/dashboard/project/_/database/roles) after you select your project. We also have a separate guide on too many connections [here](https://supabase.com/docs/guides/platform/troubleshooting#too-many-open-connections).

Reviewer 2: ```suggestion
- **Too many connections:** Every connection to your database consumes memory. You can check the number of active connections under [Database Roles](https://supabase.com/dashboard/project/_/database/roles) after you select your project. Read our guide on [too many open connections](https://supabase.com/docs/guides/platform/troubleshooting#too-many-open-connections).
```

> - **Extensions:** Some extensions such as timescale or pg_cron can use a lot of memory. It can also add up when you have too many extensions running.

Reviewer: ```suggestion
- **Extensions:** Some extensions such as timescale or pg_cron can use a lot of memory. It can also add up when you have too many extensions running. You can manage your database extensions [here](https://app.supabase.com/project/_/database/extensions).
```

> - **Extensions:** Some extensions such as `timescaledb` or `pg_cron` can use a lot of memory. It can also add up when you have too many extensions running. You can manage your database extensions [here](https://app.supabase.com/project/_/database/extensions).

Reviewer 2: ```suggestion
- **Extensions:** Some extensions such as `timescaledb` or `pg_cron` can use a lot of memory. It can also add up when you have too many extensions running. You can manage your database extensions in the dashboard under [Extensions](https://app.supabase.com/project/_/database/extensions).
```

## How to fix your memory issues

> 1. **Upgrade your compute:** You can get a Compute Add-on for your project. Follow [this link](https://app.supabase.com/project/_/settings/billing/update/pro) and select your project to see your upgrade options.

Reviewer 2: ```suggestion
1. **Upgrade your compute:** You can get a Compute Add-on for your project. See your [upgrade options](https://app.supabase.com/project/_/settings/billing/update/pro) by selecting your project.
```

> 1. **Upgrade your compute:** You can get a Compute Add-on for your project. Follow [this link](https://app.supabase.com/project/_/settings/billing/update/pro) (follow [this link](https://app.supabase.com/project/_/settings/billing/update/team) if for the Team plan and this link for the Enterprise plan) and select your project to see your upgrade options.

Reviewer: I just made a change that it will automatically redirect to the correct plan (https://github.com/supabase/supabase/pull/15036)

So you can just link to  https://app.supabase.com/project/_/settings/billing/update/pro
```suggestion
1. **Upgrade your compute:** You can get a Compute Add-on for your project. Follow [this link](https://app.supabase.com/project/_/settings/billing/update/pro) and select your project to see your upgrade options.
```

Author: ah great thanks!
